File: 417-PacificAtlanticWaterFlow/417-PacificAtlanticWaterFlow.cpp
// Last updated: 22/07/2025, 13:53:22
class Solution {
public:
    int n, m;
    int drow[4] = {-1,0,1,0};
    int dcol[4] = {0,1,0,-1};
    void dfs(int row, int col, vector<vector<int>>& heights, vector<vector<bool>>& ocean){
        ocean[row][col] = true;
        for(int i = 0; i<4; i++){
            int nrow = row + drow[i];
            int ncol = col + dcol[i];
            if(nrow >= 0 && nrow <n && ncol >=0 && ncol < m && !ocean[nrow][ncol] && heights[nrow][ncol] >= heights[row][col]){
                dfs(nrow, ncol, heights, ocean);
            }
        }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        n = heights.size();
        m = heights[0].size();
        vector<vector<bool>> pac(n, vector<bool>(m,false));
        vector<vector<bool>> atl(n, vector<bool>(m, false));
        for(int i = 0; i<n;i++){
            dfs(i,0,heights,pac);
            dfs(i,m-1,heights, atl);
        }
        for(int i = 0; i<m; i++){
            dfs(0,i, heights, pac);
            dfs(n-1, i , heights, atl);
        }
        vector<vector<int>> res;
        for(int i = 0; i<n; i++){
            for(int j = 0; j < m; j++){
                if(pac[i][j] && atl[i][j]){
                   res.push_back({i,j});
                }
            }
        }
        return res;
    }
};

File: 417-PacificAtlanticWaterFlow/417-PacificAtlanticWaterFlow.py
# Last updated: 01/10/2025, 00:12:55
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows, cols = len(heights), len(heights[0])
        pacific,atlantic = set(),set()

        def dfs(r,c,visited,prevHeight):
            if ((r,c) in visited or r<0 or c<0 or r>=rows or c >= cols or heights[r][c] < prevHeight):
                return None
            visited.add((r,c))
            dfs(r+1,c,visited,heights[r][c])
            dfs(r-1,c,visited,heights[r][c])
            dfs(r,c+1,visited,heights[r][c])
            dfs(r,c-1, visited, heights[r][c])
        
        for c in range(cols):
            dfs(0,c,pacific, heights[0][c])
            dfs(rows-1,c,atlantic,heights[rows-1][c])
        for r in range(rows):
            dfs(r,0,pacific,heights[r][0])
            dfs(r,cols-1,atlantic, heights[r][cols-1])
        result = list(pacific & atlantic)
        return result